### ISSUE TITLE
Receiving playground-service-worker-proxy.html 404 (Not Found) when using playground with webpack

### ISSUE THREAD

**Issue by codecycletechnologies** [PolymerLabs/playground-elements#131]
Hello, 

We are trying to setup playground-elements using webpack and typescript.
Sample Repo - `https://github.com/codecycletechnologies/playground-elements-webpack`

It seems playground-elements is using playground-service-worker-proxy.html internally from its package. However when we try to serve with web-dev-server of webpack, we are receiving error `"GET http://localhost:8080/[object%20Module]/playground-service-worker-proxy.html 404 (Not Found)". ` 

What does [object%20Module] means? On which path it is expecting to have playground-service-worker-proxy file? 
Can you please help!

Thanks

**Comment by aomarks** [comment]
Hi @codecycletechnologies 

Webpack seems to be transforming [`import.meta.url`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import.meta) in a broken way that isn't compatible with standard JavaScript semantics. This particular case seems like an especially bad bug -- it's replacing `import.meta.url` with some code that produces an object, instead of a string.

I'd generally recommend using Rollup instead of Webpack, because Rollup supports emitting JavaScript modules, whereas Webpack only supports emitting legacy JavaScript code. There's no down-side to emitting modules when using Playground, because there are no browsers that support Service Workers but don't support modules.

I think you *can* get it to work with Webpack, though. Can you try this?

1. Update the `CopyPlugin` section of your `webpack.config.js` like this:

```js
new CopyPlugin({
  patterns: [
    { from: './src/greetings/', to: './greetings/' },
    { from: './node_modules/playground-elements/playground-service-worker-proxy.html', to: './playground-service-worker-proxy.html' },
    { from: './node_modules/playground-elements/playground-service-worker.js', to: './playground-service-worker.js' },
  ],
})
```

2. Update your `index.ts` like this:
```js
<playground-ide line-numbers resizable editable-file-system project-src="./greetings/say-hello.json" sandbox-base-url="/"> </playground-ide>
```

**Comment by e111077** [comment]
This is a confirmed fix in the given reproduction

https://github.com/codecycletechnologies/playground-elements-webpack

**Comment by thw0rted** [comment]
Thanks for the details, @aomarks. I'm actually not a Playground user but was trying to help the OP with the webpack side of things.  (Not a dev there either, FWIW.) I'm just looking at how Playground is consuming `import.meta.url`.  Webpack is trying to emulate browser behavior when `import.meta.url` is encountered, typically by replacing the defined URL with its own module system.  So, if you have `const worker = new Worker(new URL("./worker.js", import.meta.url))`, it should determine through static analysis that the `new URL(...)` part can be replaced by an in-memory script module, maybe encoded as a data-URL.  The [docs for this feature](https://webpack.js.org/guides/asset-modules/#url-assets) are still a little light.

Your usage of `import.meta.url` must not match what the static analysis expects.  I found [this reference](https://github.com/PolymerLabs/playground-elements/blob/main/src/playground-service-worker-proxy.ts#L45) which is just calling `new URL(url, import.meta.url).href` -- the hard part is that the `url` parameter is an arbitrary string, unknown at compile-time.  I know [the Webpack team has asked for use cases](https://github.com/webpack/webpack/issues/7353) to make their tooling smarter.  Maybe this is a good one?

**Comment by alexander-akait** [comment]
@aomarks 

> Webpack seems to be transforming import.meta.url in a broken way that isn't compatible with standard JavaScript semantics. This particular case seems like an especially bad bug -- it's replacing import.meta.url with some code that produces an object, instead of a string.

What is broken/wrong? Please clarify. We use the same semantic as js

> I'd generally recommend using Rollup instead of Webpack, because Rollup supports emitting JavaScript modules, whereas Webpack only supports emitting legacy JavaScript code.

You can enable module output, don't see any problems here.

Please describe the problem(s). when somebody say that something is working incorrectly, usually give an example that it does not work correctly/broken, thanks

**Comment by aomarks** [comment]
Thanks for responding @thw0rted and @alexander-akait 

Apologies if I sounded dismissive and wasn't specific about the issue in my original comment. I appreciate that down-leveling is complicated and it may only be practical to support particular usage patterns.

I think the summary here is that while Webpack does support e.g. `new Worker(new URL('./worker.js', import.meta.url))`, it's not able to follow indirect references to `import.meta.url`, like in [this code](https://github.com/PolymerLabs/playground-elements/blob/80afd98ef0e548251e06496808625cfe8ce81f25/src/playground-project.ts#L160). It's understandable that supporting this kind of indirection is difficult, but I was a bit surprised by what Webpack does instead, because it seems to either construct invalid URLs, or use the `file://` protocol, depending on the particular phrasing (details below).

The reason this project [assigns](https://github.com/PolymerLabs/playground-elements/blob/80afd98ef0e548251e06496808625cfe8ce81f25/src/playground-project.ts#L160) `import.meta.url` to an intermediate property in this way is that it represents a base path from which two URLs are later constructed. So, it defaults to "wherever this module was imported from" (i.e. the NPM package directory) using `new URL('.', import.meta.url)`, but can be overridden by the user (in fact as of https://github.com/PolymerLabs/playground-elements/pull/137 it defaults to a different URL entirely, for a different security-related reason, which also happens to fix this issue).

---

> What is broken/wrong? Please clarify. We use the same semantic as js

When executed directly as a JavaScript module, the following code creates a URL object representing the containing path of the module:

```js
// Original JS
const base = new URL('.', import.meta.url);
console.log(base.href); // http://localhost:8080/
```

However, Webpack converts this `import.meta.url` to a `__webpack_require__` function call. This function returns an object, so it ends up constructing a URL with a stringified object in it:

```js
// Webpack output
const base = new URL(/* asset import */ __webpack_require__(/*! . */ "./node_modules/playground-elements/playground-ide.js"), __webpack_require__.b);
console.log(base.href); // http://localhost:8080/[object%20Module]
```

I also tried the following code, which is equivalent in JavaScript, but produces something very different from Webpack:

```js
// Original JS
const importUrl = import.meta.url;
const base = new URL('.', importUrl);
console.log(base.href); // http://localhost:8080/
```

```js
// Webpack output
const importUrl = "file:///home/aomarks/scratch/cycle/playground-elements-webpack/node_modules/playground-elements/playground-project.js";
```

I was surprised to see a `file://` URL here, since that protocol is very restricted (e.g. can't be used for worker scripts or fetch calls), and Webpack is serving the script itself over HTTP, not `file://`.

I should also point out that I'm directly modifying a script in a checkout of the original poster's `node_modules/` directory here. I noticed that when I reproduce similar code in the main sources of a fresh Webpack project, I get an error. I'm not sure if this is because Webpack handles errors in `node_modules/` dependencies differently, or if it's some other configuration difference.

```
Module not found: Error: Can't resolve '.' in '/home/aomarks/scratch/webpack/src'
```

---

> You can enable module output, don't see any problems here.

I looked for an option to emit JS modules at https://webpack.js.org/configuration/target/, but I don't see it there. And https://github.com/webpack/webpack/issues/2933#issuecomment-774253975 seems to suggest outputting modules is still a work in progress? Is there an option I missed?

**Comment by aomarks** [comment]
@codecycletechnologies We just released v0.9.2 which means Webpack should now work without any special configuration. See https://github.com/PolymerLabs/playground-elements#bundling for documentation and a link to an example.

(The reason for the change in this release was actually primarily security related, see https://github.com/PolymerLabs/playground-elements/pull/137, but it had the nice side effect of side-stepping the Webpack issue here too).

**Comment by codecycletechnologies** [comment]
@aomarks Thanks for the fix... after upgrading to v0.9.2, it is working fine.
